[RPD-2865] View-ViewReactor연결

ehdudgnsTW opened the pull request:
<!--
  제목은 `[지라이슈번호] 작업 내용 요약`으로 작성해주세요.
-->

## 설명
### 작업 내용
기존에 viewController에서 더미 데이터를 가지고 와서 뷰에 그냥 띄워주는 방식으로 구현 한 부분을
repository에서 임의의 목업 데이터를 생성하였습니다
repository에서 생성된 목업 데이터는 mainviewreactor를 통해서 mainviewcontroller에 보내주는 형식으로 구현 하였습니다.
즐겨찾기 기능을 구현 하였습니다. 별모양 버튼을 눌렀을경우 추가 또는 해제가 가능합니다.
즐겨찾기 데이터가 없을경우에도 추가가 가능합니다.


### 스크린샷


## 기타
Repository파일에서 목업 데이터를 생성하였습니다.
MainViewReactor에서 즐겨찾기 및 장소 검색 기능을 구현 했습니다.
MainViewController에서 cell을 구성하고 구현 하였습니다.
LocationDataTableViewCell, FavoriteDataTableViewCell 안에 있는 버튼을 MainViewReactor랑 연결을 하였습니다.

## 참고 링크

hyeonsikTW reviewed (commented):
즐겨찾기 등록 후 앱을 종료하고 다시 열면 즐겨찾기가 등록 되어있지 않은데 아직 저장하는 부분은 구현이 안된거죠?

ehdudgnsTW commented:
> 즐겨찾기 등록 후 앱을 종료하고 다시 열면 즐겨찾기가 등록 되어있지 않은데 아직 저장하는 부분은 구현이 안된거죠?

그 부분에 있어서는 아직 구현을 하지 못하였습니다. 구현 후에 다시 올리도록 하겠습니다.

ehdudgnsTW committed:
Remove TableViewCellReactor file

TabelViewCellReactor 파일을 사용하지 않아서 삭제 하였습니다.

PRD-2865

ehdudgnsTW committed:
Add 기상청41_단기예보_조회서비스_오픈API활용가이드.csv파일 추가

공공데이터 api통신을 하기위해서 지역명에 따라 격자 데이터들이 필요하여서 추가를 하였습니다.

PRD-2865

ehdudgnsTW committed:
Fix FavoriteDataModel in cityTemperature from constant to variable

cityTemperature가 기존에는 상수로 선언이 되어있었는데 이 변수는 값이 변경이 되어야 하므로
let에서 var로 변경을 하였습니다.

RPD-2865

ehdudgnsTW committed:
Fix FavorietDataModel file

api를 받아오고 데이터들을 띄우기 위해서 모델의 변수명과
필요한 데이터들이 많아져서 수정을 하였습니다.

RPD-2865

ehdudgnsTW committed:
Add Repository file

api통신을 하기전에 목업 데이터를 생성하기 위해서 MockRepository class를 생성하였고
Repository class는 실제로 api통신을 하기위해서 만들어 놓은 클래스 입니다.
그리고 이둘은 같은 형식을 가져야 한다고 생각을 해서 프로토콜을 만들어서 진행하였습니다.

RPD-2865

ehdudgnsTW committed:
Add reactor code for Reactorkit in MainViewController

MainViewController에다가 MainViewReactor를 연결하기 위해서 View를 extension하고
프로토콜에서 필수적으로 선언해야 하는 것들을 선언하는 코드를 추가 하였습니다.

RPD-2865

ehdudgnsTW commented:
작은 단위로 커밋을 위해서 임시적으로 pull request를 닫았습니다. 커밋 완료 하면 다시 재오픈 하도록 하겠습니다.

ehdudgnsTW committed:
Refactor MainViewController and Refactor MainViewReactor

MainViewController에 init함수를 추가 하면서 뷰리엑터를 연결 하였습니다.

-MainViewController 수정
-AppDelegate 와 SceneDelegate에 ViewController 선언 할때 파라미터 추가
-MainViewReactor 수정
-Repositiry 파일에서 마지막 데이터가 안나오는 이슈가 있어서 범위를 변경

RPD-2865

ehdudgnsTW committed:
Implement MainViewReactor action search

MainViewReactor에 검색 기능의 액션과 검색결과에 맞는 데이터들을 보내주도록 구현하였습니다.

RPD-2865

ehdudgnsTW committed:
Implement MainViewController view bind

MainViewController에 있는 bind 부분을 reactor랑 연결을 한뒤에 view에 테이블 뷰로 뿌려주는 작업을 하였습니다.

RPD-2865

ehdudgnsTW committed:
Make filter search cell and favorite cell

지금 현재 지역명을 searching할때와 searching을 안할때의 화면을 구분하여 구현 하였습니다.

-MainViewController부분에서 bind부분의 코드 추가

RPD-2865

ehdudgnsTW committed:
Add Set Button Image according to isFavorite

현재 즐겨찾기 되어있는지 안되어있는지에 따라서 버튼의 이미지를 변경시키는 코드를 전역적으로 구현을 하였습니다.

-Common/ImageAsset 파일 수정

RPD-[2865]

ehdudgnsTW committed:
Use imageSetting function in ImageAsset

ImageAsset에 구현된 imageSetting을 사용하여 즐겨찾기 버튼의 이미지를 적용 시켰습니다.
또한 이에 따라 필요한 데이터들이 달라 파라미터를 수정 하였습니다.

-FavoriteDataTableViewCell의 button image이름 수정
-LocationDataTableViewCell의 configureSearchingView 수정
-MainViewController의 cell configure부분의 파라미터 수정

RPD-2865

ehdudgnsTW committed:
ADD search start and search end status

searchcontroller를 통해 서칭의 시작 부분과 서칭의 끝나는 부분의 액션을 추가 하였습니다.

-MainViewController의 bind 수정

RPD-2865

ehdudgnsTW committed:
ADD Favorited function in reactor

즐겨찾기 버튼을 눌렀을때 즐겨찾기의 상태를 변경시켜 줄 수 있는 액션을 추가 하였습니다

-MainViewReactor의 Action을 추가

RPD-2865

ehdudgnsTW committed:
ADD Favorite button function connection for each cell

각각 서칭중일때이나 즐겨찾기 화면에서 즐겨찾기의 상태를 변경 할 수 있도록 수정하였습니다.

-FavorietDataTableViewCell,LocationDataTableViewCell의 bind및 reactor disposebag 추가
-MainViewController에서 cell의 reactor 연결

RPD-2865

ehdudgnsTW committed:
Fix button not being tap

셀의 버튼이 잘눌리지 않는 현상이 있어서 수정 하였습니다.

RPD-2865

ehdudgnsTW committed:
ADD favoriteData action in MainViewReactor

즐겨찾기만 가져올 수 있는 액션을 추가하였습니다.

RPD-2865

ehdudgnsTW committed:
Fix reactor action when searchbar textendediting

서치바에서 서칭이 끝났을때 즐겨찾기 목록만 불러 오도록 수정 하였습니다.

RPD-2865

nexmondTW commented on TwinnyAssignmentiOSProject/Common/ImageAsset.swift:
```diff
@@ -8,6 +8,15 @@
+        else {
+            self.setImage(SizeStyle.resizeImage(image: UIImage(named: "star_black"), 20, 20), for: .normal)
+        }
+    }
```
즐겨찾기 전용 이미지 설정 함수입니다. 일반화 할 필요가 있습니다.

nexmondTW commented on TwinnyAssignmentiOSProject/Repository/Repository.swift:
```diff
@@ -0,0 +1,69 @@
+            if let dataArr = dataEncoded?.components(separatedBy: "\n").map({
+                $0.components(separatedBy: "\",\"")
+            }) {
+                for i in 1..<dataArr.count-1 {
```
`dataArr.count`가 아닌 `dataArr.count-1`인 이유가 있나요?

ehdudgnsTW replied:
dataArr의 마지막 값이 아무값도 존재 하지 않아서 마지막값을 제외하고 for문을 돌렸습니다.

nexmondTW commented on TwinnyAssignmentiOSProject/VIews/Cells/FavoriteDataTableViewCell.swift:
```diff
@@ -25,24 +30,31 @@ class FavoriteDataTableViewCell: UITableViewCell {
+        dataModel = favoriteContents
+        self.contentView.addSubview(cityTemperature)
+        self.contentView.addSubview(cityName)
+        self.contentView.addSubview(favoriteButton)
```
셀에 데이터를 바인딩할 때 마다 레이아웃을 불필요하게 다시 잡고 있습니다.

nexmondTW commented on TwinnyAssignmentiOSProject/VIews/MainViewController.swift:
```diff
@@ -67,6 +118,12 @@ class MainViewController: UIViewController {
     }
     
 }
+extension Reactive where Base: MainViewController {
+    var viewDidLoad: ControlEvent<Void> {
+        let source = self.methodInvoked(#selector(Base.viewDidLoad)).map { _ in }
+        return ControlEvent(events: source)
+    }
+}
```
일반화하여 개별 파일로 분리하는 것을 권장합니다.

nexmondTW commented on TwinnyAssignmentiOSProject/ViewModels/MainViewReactor.swift:
```diff
@@ -8,28 +8,100 @@
+    func mutate(action: Action) -> Observable<Mutation> {
+        switch action {
+        case .setTotalData:
+            repository.getCSVFileToData().subscribe {
+                data in
+                self.totalData = data
+            }.disposed(by: DisposeBag())
+            let favoriteDatas = totalData.filter {
+                $0.isFavoriet
+            }
+            return .just(Mutation.favoriteData(favoriteDatas))
```
totalData를 Rx 방식으로 가져오는 이유가 있나요?

nexmondTW commented on TwinnyAssignmentiOSProject/ViewModels/MainViewReactor.swift:
```diff
@@ -8,28 +8,100 @@
+    func reduce(state: State, mutation: Mutation) -> State {
+        var state = state
+        switch mutation {
+        case .favoriteData(let array):
+            state.isSearching = false
+            state.filterData = array
+            return state
+        case .searchingData(let array):
+            state.isSearching = true
+            state.filterData = array
+            return state
+        }
```
같은 형태인데, 다른 Mutation으로 받는 이유가 있나요?

nexmondTW reviewed (commented):
- 처음 앱 실행 시 모든 위치가 즐겨찾기로 되어 있습니다.
- 검색 중 즐겨찾기 아이콘 선택시 목록이 비정상적으로 변경됩니다.
- 메인 즐겨찾기 셀의 즐겨찾기 버튼 영역이 너무 넓습니다. 정사각형 크기로 줄이세요.

ehdudgnsTW committed:
Fix data where favoriteDatas

초기에 즐겨찾기등록 되어서 나오는 부분을 수정 하였습니다.

RPD-2865

ehdudgnsTW committed:
Refactor MainViewReactor

불필요하게 나누어져있던 Mutation을 하나로 합치고 상태에 따라서 다른값을 반환하게 변경 하였습니다.
또한 setTotalData도 불필요해보여서 삭제를 하였습니다.
totalData의 경우 rx로 받을 필요가 없어보여서 reactor가 init할때 totalData를 세팅하도록 변경하였습니다.

RPD-2865

ehdudgnsTW committed:
Fix button size in FavoriteDataCell

기존의 즐겨찾기 버튼의 너비나 크기가 커서 작게 변경 하였습니다.

RPD-2865

ehdudgnsTW committed:
Fix configure function in FavoriteDataTableViewCell and LocationDataTableViewCell

각 셀에 바인딩 할때마다 레이아웃을 다시 잡지 않게 하기 위해서 cell을 init할때 layout을 설정 하도록 수정 하였습니다.

RPD-2865

ehdudgnsTW committed:
Fix imageSetting function from extension to general

기존에 uibutton안에 하나의 함수로 만들어서 사용했던 imageSetting을 일반화 하여 버튼을 넣어서 이미지를 세팅하게 수정 하였습니다.

RPD-2865

ehdudgnsTW committed:
Move extension Reactor code to new file ExtensionReactorViewController

기존에 MainViewcontroller에 있는 extension Reactive부분을 새로운 파일로 만들어서 옮겼습니다.
또한 기존의 base의 경우 MainViewController로만 지정을 했는데 이제는 모든 ViewController에서 사용할수 있게 바꾸었습니다.

RPD-2865

ehdudgnsTW committed:
Rename imageSetting to favoriteStateStarImageSetting

기존의 의미가 모호했던 함수의 이름을 변경하여서 즐겨찾기 상태에서 이미지를 세팅하는 함수라는 것을 알 수 있게 변경 하였습니다.

RPD-2865

ehdudgnsTW committed:
Fix FavDataCell cityName label and button constraints

기존에 라벨이 뷰에서 깨진부분을 수정 하였고 버튼의 이미지가 오른쪽으로 쏠려있는것을 해결하였습니다.

RPD-2865

nexmondTW commented:
수고하셨습니다.
피드백이 모두 반영되어 합칩니다.
